### ISSUE TITLE
GPU utilization issue

### ISSUE THREAD

**Issue by Alex18947** [ChipsSpectre/chipstorch-yolov3#1]
Hi,

this repository has the same issue as the orignal repository. When doing inference on GPU, the GPU cannot be fully utilized while CPU usage is high. To reproduce create a loop around the forward call like this:

`while(1) {
auto result = forward(img_var);
}`

You will see that GPU load goes only up to 60 percent and CPU usage is still high.

Expected result: GPU should be fully utilized, CPU usage should stay small. So there is a bottleneck somewhere with using many CPU cycles. As a result, the total throughput is about 30 percent lower compared to original Darknet Yolov3 which can fully utilize the GPU.

I am testing on Windows and with latest libtorch.

A.

**Comment by ChipsSpectre** [comment]
Hey

Thank you for pointing that out!

I looked into the code and I think this issue occurs because the different modules/layers are
executed in a sequence. So after each module, it has to be synchronized with CPU.

The reason for this approach are the layers that depend on more than the directly
previous layer, which make it harder to implement yolov3 using the `forward()`function 
of the `Sequential` class.

In order to change this, delete the custom forward() function and create a container that
stores all previous activations. Then, you make this container globally accessible and let
the problematic layers access this container (which is filled by each layer), so that the default `Sequential` `forward()` implementation can be used.

Contributions are welcome!

Best regards
Max

**Comment by Alex18947** [comment]
Yes, I have the exact same feeling about what the cause is. I am not quite sure, however, if the solution is that easy as you described. Since the shortcut and route layers need the output of some other than only the previous layers, a CPU sync will be needed at least before each such layer. Or do you think otherwise?

A.

**Comment by Alex18947** [comment]
So after making some heavy changes to the code - all layers now save the output to a global container and after re-implementing the forward() function so that it only contains one forward call for the very first Sequential (so no loop any more), the result is still the same :(, GPU goes only up to 70 percent and CPU is high, some sync is still happening...

Any other ideas?

A.

**Comment by ChipsSpectre** [comment]
I assume that a CPU sync is necessary before each of the shortcut layers.

I have another idea - maybe the implementation of forward() in the `DetectionLayer` triggers CPU execution. If some tensors there do not reside on the CPU, then image-based operations would be done on CPU instead of GPU what would explain your high CPU usage.

**Comment by Alex18947** [comment]
Hi, as for your suggestions (after spending some more time with that):

- I don't believe that a CPU sync is necessary for a shortcut layer, it merely references some other GPU tensors. And even if a sync happened here, this would only make the forward() slower and wouldn't, IMHO, increase CPU usage.

- I went through the DetectionLayer in detail, it uses GPU tensors only with one exception (see [here](https://github.com/ChipsSpectre/chipstorch-yolov3/blob/master/include/chipsYolo/DetectionLayer.hpp#L48)). But this is a short piece of code and since there are only 3 such layers in total, this can never result in such high CPU usage.

The weird thing is that compared to original Darknet the detection times are comparable with the only difference that the libtorch version increases CPU consumption.

I am out of ideas - at least for now ;).

A.

**Comment by Alex18947** [comment]
Well, just to update you on this, I already (hope to) know what's happening here (and it is actually quite simple). When running on GPU, the GPU calls are extremely fast, they return so fast that when testing maximum throughput by running the detections in a loop, the CPU overhead of the forward() method and of the various tensors ops becomes prominent and the CPU thread running the loop kills one CPU core even before GPU can be fully utilized - this is the increase in CPU usage you can see. Simple, right? ;)

This is also consistent with the fact that if you increase the image size to 608 or 832, one CPU core is still killed (at least in my testing environment), but the GPU load increases from 70 to 80 or 90 percent, because the GPU calls need more time now.

Finally, how does Darknet eat up the the entire GPU while being low in CPU consumption? Well, look at the source code. Darknet moves along the layer infrastructure with a simple C loop using some GPU pointers. This is so fast, that even the GPU operations will not outperform this CPU loop. Result will be a full GPU utilization and low CPU. OTOH, libtorch with its tensors uses a lot of virtuals calls with many VTABLE lookups, the CPU overhead is much bigger. There is some ongoing work to make the base objects perform better (see [here](https://github.com/pytorch/pytorch/issues/13049) for example). I strongly believe that this will boost the libtorch Yolov3 as well.

Let me know your thoughts. 

A.

**Comment by ChipsSpectre** [comment]
Hey

This explanation makes really sense to me. Since the GPU load increases with the image size, the CPU consumption seems to depend on the cost of the tensor operations that are done by libtorch. I hope that this will be fixed soon by the PyTorch developers.

Thank you really much for your effort!

Best regards
ChipsSpectre

**Comment by Sephirot1st** [comment]
> Well, just to update you on this, I already (hope to) know what's happening here (and it is actually quite simple). When running on GPU, the GPU calls are extremely fast, they return so fast that when testing maximum throughput by running the detections in a loop, the CPU overhead of the forward() method and of the various tensors ops becomes prominent and the CPU thread running the loop kills one CPU core even before GPU can be fully utilized - this is the increase in CPU usage you can see. Simple, right? ;)
> 
> This is also consistent with the fact that if you increase the image size to 608 or 832, one CPU core is still killed (at least in my testing environment), but the GPU load increases from 70 to 80 or 90 percent, because the GPU calls need more time now.
> 
> Finally, how does Darknet eat up the the entire GPU while being low in CPU consumption? Well, look at the source code. Darknet moves along the layer infrastructure with a simple C loop using some GPU pointers. This is so fast, that even the GPU operations will not outperform this CPU loop. Result will be a full GPU utilization and low CPU. OTOH, libtorch with its tensors uses a lot of virtuals calls with many VTABLE lookups, the CPU overhead is much bigger. There is some ongoing work to make the base objects perform better (see [here](https://github.com/pytorch/pytorch/issues/13049) for example). I strongly believe that this will boost the libtorch Yolov3 as well.
> 
> Let me know your thoughts.
> 
> A.

Hello, I am using [walktree's implementation](https://github.com/walktree/libtorch-yolov3) of yolov3 in libtorch and make some changes to it, now the CPU usage is very high. Could you provide your solution to this problem?
Thanks.

**Comment by Sephirot1st** [comment]
> So after making some heavy changes to the code - all layers now save the output to a global container and after re-implementing the forward() function so that it only contains one forward call for the very first Sequential (so no loop any more), the result is still the same :(, GPU goes only up to 70 percent and CPU is high, some sync is still happening...
> 
> Any other ideas?
> 
> A.

How do you deal with the shortcut and route layers in yolov3? How to merge the modules in just one forward call? Could you share your ideas?

**Comment by ChipsSpectre** [comment]
Hey :smile: 

Currently, the forward() call is done by the Yolo class (*Yolo.hpp*) and has the following method signature

`torch::Tensor forward(torch::Tensor x)`

(takes the image tensor and returns the bounding boxes).

So, currently the `forward()` method already is done in one method.
It deals with the _shortcut_ and _route_ layers by creating a vector of intermediate results, and referring to them when needed.
The numbers in the darknet .cfg file at these layers (said simplified) just refer to which previous layer to use for its operation.

As **Alex18947** pointed out, the main point for high CPU usage is the large number of calls PyTorch does to handle its tensors. The classical _darknet_ framework just iterates over a C array and increments a pointer, which results in lower CPU use.

 Best regards
